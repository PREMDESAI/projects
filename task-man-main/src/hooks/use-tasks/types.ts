export type TaskStatus = 'todo' | 'in_progress' | 'in_review' | 'done';

export interface Task {
  id: string;
  title: string;
  description?: string;
  status: TaskStatus;
}

export interface TaskContextValues {
  isLoading: boolean;
  // Data Layer
  tasksData: Task[];
  // View Layer
  tasksView: Task[];
  statuses: TaskStatus[];

  addTask: (task: Omit<Task, 'id'>) => Promise<void>;
  updateTask: (id: Task['id'], updateTask: Omit<Task, 'id'>) => Promise<void>;
  deleteTask: (id: Task['id']) => Promise<void>;
  filterTasks: (parameters: { text?: string; taskStatus?: TaskStatus }) => void;
}
