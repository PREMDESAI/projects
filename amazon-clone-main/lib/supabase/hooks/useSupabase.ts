import { useState } from "react"
import { supabase } from "../products";

export const useSupabase = () => { 
    const [products, setProducts] = useState<any>([]);
    const [filterData, setFilterData] = useState<any>([]);
    const [singleProduct, setSingleProduct] = useState<any>([]);
    const [mensProduct, setMensProduct] = useState<any>([]);
    const [womensProduct, setWomensProduct] = useState<any>([]);

    const getDataFromSupabase = async () => {
        let {data, error} = await supabase.from('product').select("*");
        if(data){
            setProducts(data); 
        }
        if(error){
            console.log(error);
        }
    }
    const getFilteredData = async (query:string) => {
        let {data, error} = await supabase.from('product').select("*").or(`title.ilike.%${query}%, description.ilike.%${query}%, category.ilike.%${query}%`); // cloth
        if(data){
            setFilterData(data); 
        }
        if(error){
            console.log(error);
        }
    }

    const getSingleProduct = async (id:number)=>{
        let {data, error} = await supabase.from('product').select('*').eq('id', id);
        if(data){
            setSingleProduct(data);
        }
        if(error){
            console.log(error);
        }
    }
    const getMensClothing = async () => {
        let {data, error} = await supabase.from('product').select('*').ilike('category', `men's clothing`);
        if(data){
            setMensProduct(data);
        }
        if(error){
            console.log(error);
            
        }
    }
    const getWomensClothing = async () => {
        let {data, error} = await supabase.from('product').select('*').ilike('category', `women's clothing`);
        if(data){
            setWomensProduct(data);
        }
        if(error){
            console.log(error);
            
        }
    }

    
    return {
        products, 
        getDataFromSupabase,
        filterData,
        getFilteredData,
        singleProduct,
        getSingleProduct,
        mensProduct,
        getMensClothing,
        womensProduct,
        getWomensClothing
    };
}