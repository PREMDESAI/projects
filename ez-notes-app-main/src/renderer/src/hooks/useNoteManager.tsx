// src/hooks/useNoteManager.ts
import { useState, useRef, useEffect } from 'react'
import { v4 as uuidv4 } from 'uuid'
import debounce from 'lodash.debounce'
import { Note } from '../global'
import { doc, setDoc } from 'firebase/firestore'
import { db } from '../firebase'

export function useNoteManager() {
  const [title, setTitle] = useState('')
  const [content, setContent] = useState('')
  const [currentNoteId, setCurrentNoteId] = useState<string | null>(null)
  const [notes, setNotes] = useState<Note[]>([])
  // const [notesSummary, setNotesSummary] = useState<Note[]>([])

  const titleRef = useRef(title)
  const contentRef = useRef(content)
  const currentNoteIdRef = useRef(currentNoteId)

  useEffect(() => {
    titleRef.current = title
  }, [title])

  useEffect(() => {
    contentRef.current = content
  }, [content])

  useEffect(() => {
    currentNoteIdRef.current = currentNoteId
  }, [currentNoteId])

  useEffect(() => {
    // Load existing notes on mount
    const loadNotes = async () => {
      try {
        const loadedNotes = await window.electronAPI.readAllNotes()
        setNotes(loadedNotes)
      } catch (error) {
        console.error('Error loading notes:', error)
      }
    }

    loadNotes()
  }, [])

  const handleSaveNote = async () => {
    const note = {
      id: currentNoteIdRef.current || uuidv4(),
      title: titleRef.current,
      content: contentRef.current
    }
    saveNoteContent(note.id, note.title, note.content)

    if (currentNoteIdRef.current) {
      // Update existing note
      await window.electronAPI.updateNote(note)
    } else {
      // Create new note
      await window.electronAPI.createNote(note)
      setCurrentNoteId(note.id)
      currentNoteIdRef.current = note.id
      setNotes((prevNotes) => [
        ...prevNotes,
        { id: note.id, title: note.title, content: note.content }
      ])
    }
  }

  const saveNoteContent = async (noteId, title, content) => {
    try {
      await setDoc(doc(db, 'notes', noteId), {
        title,
        content,
        lastEdited: new Date() // Optional
      })
    } catch (e) {
      console.error('Error writing document: ', e)
    }
  }

  const debouncedSave = useRef(
    debounce(() => {
      console.log('--FIREBASE CALL--')

      handleSaveNote()
    }, 1000)
  ).current

  useEffect(() => {
    return () => {
      debouncedSave.cancel()
    }
  }, [debouncedSave])

  const handleTitleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    setTitle(e.target.value)
    debouncedSave()
  }

  const handleContentChange = (newContent: string) => {
    setContent(newContent)
    debouncedSave()
  }

  const handleSelectNote = async (id: string) => {
    try {
      const note = await window.electronAPI.readNote(id)
      if (note) {
        setTitle(note.title)
        setContent(note.content)
        setCurrentNoteId(note.id)
        currentNoteIdRef.current = note.id
      }
    } catch (error) {
      console.error('Error loading note:', error)
    }
  }

  return {
    title,
    content,
    notes,
    handleTitleChange,
    handleContentChange,
    handleSelectNote
  }
}
