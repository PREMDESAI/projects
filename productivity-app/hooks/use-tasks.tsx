// hooks/useTasks.ts
import { useState, useEffect } from 'react';
import { SelectTask } from '../../types'; // Adjust the import path as necessary

interface UseTasksReturn {
  tasks: SelectTask[];
  loading: boolean;
  error: string | null;
  refetch: () => void;
}

export const useTasks = (params: string): UseTasksReturn => {
  const [tasks, setTasks] = useState<SelectTask[]>([]);
  const [loading, setLoading] = useState<boolean>(true);
  const [error, setError] = useState<string | null>(null);

  const fetchTasks = async () => {
    const API_BASE_URL = process.env.NEXT_PUBLIC_API_BASE_URL;
    setLoading(true);
    setError(null);
    let url = `${API_BASE_URL}/tasks`;
    if (params) {
      url += `?params=${params.toString()}`;
    }
    try {
      const res = await fetch(`${url}`, {
        cache: 'no-store',
      });
      if (!res.ok) {
        throw new Error('Failed to fetch tasks.');
      }
      const data = await res.json();
      console.log(data.tasks);

      setTasks(data.tasks);
    } catch (err: any) {
      console.error('Error fetching tasks:', err);
      setError(err.message || 'An error occurred.');
    } finally {
      setLoading(false);
    }
  };

  useEffect(() => {
    fetchTasks();
  }, []);

  const refetch = () => {
    fetchTasks();
  };

  return { tasks, loading, error, refetch };
};
