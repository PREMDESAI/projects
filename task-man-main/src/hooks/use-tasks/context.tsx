import React from 'react';

import { Task, TaskContextValues, TaskStatus } from './types';
import { TaskRepository } from '@/models/task';
import { STATUSES } from '@/constants';

export const TasksContext = React.createContext<TaskContextValues>({
  isLoading: false,
  tasksData: [],
  tasksView: [],
  statuses: [],

  addTask: () => Promise.resolve(),
  updateTask: () => Promise.resolve(),
  deleteTask: () => Promise.resolve(),
  filterTasks: () => null,
});

export const TasksProvider = ({ children }: React.PropsWithChildren) => {
  const [tasksData, setTasksData] = React.useState<Task[]>([]);
  const [tasksView, setTasksView] = React.useState<Task[]>([]);
  const [isLoading, setIsLoading] = React.useState(false);

  React.useEffect(() => {
    setIsLoading(true);
    TaskRepository.read()
      .then(setTasksData)
      .finally(() => setIsLoading(false));
  }, []);

  React.useEffect(() => {
    setTasksView([...tasksData]);
  }, [tasksData]);

  const addTask = React.useCallback(
    async (task: Omit<Task, 'id'>) => {
      setIsLoading(true);
      setTasksData([
        ...tasksData,
        {
          ...task,
          id: `TM-${tasksData.length + 1}`,
        },
      ]);

      const updatedTasks = await TaskRepository.create(task);
      setTasksData(updatedTasks);
      setIsLoading(false);
    },
    [tasksData]
  );

  const updateTask = React.useCallback(
    async (id: string, updatedTask: Omit<Task, 'id'>) => {
      setIsLoading(true);
      const indexToUpdate = tasksData.findIndex((task) => task.id === id);

      const clonedTasks = [...tasksData];
      clonedTasks.splice(indexToUpdate, 1);

      setTasksData([
        ...clonedTasks.slice(0, indexToUpdate),
        {
          id,
          ...updatedTask,
        },
        ...clonedTasks.slice(indexToUpdate),
      ]);

      const updatedTasks = await TaskRepository.update({ id, ...updatedTask });
      setTasksData(updatedTasks);
      setIsLoading(false);
    },
    [tasksData]
  );

  const deleteTask = React.useCallback(
    async (id: Task['id']) => {
      setIsLoading(true);

      const clonedTasks = [...tasksData];
      const indexToUpdate = tasksData.findIndex((task) => task.id === id);
      clonedTasks.splice(indexToUpdate, 1);
      setTasksData(clonedTasks);

      const updatedTasks = await TaskRepository.remove(id);
      setTasksData(updatedTasks);
      setIsLoading(false);
    },
    [tasksData]
  );

  const filterTasks = React.useCallback(
    ({ taskStatus, text }: { text?: string; taskStatus?: TaskStatus }) => {
      if (!taskStatus && !text) {
        setTasksView([...tasksData]);
        return;
      }
      const sanitizedText = text?.trim().toLowerCase();
      setTasksView([
        ...tasksData.filter(({ id, title, status }) => {
          let textMatch: boolean = false;
          let statusMatch: boolean = false;

          if (sanitizedText) {
            textMatch =
              id.toLowerCase().includes(sanitizedText) ||
              title.toLowerCase().includes(sanitizedText);
          }

          if (taskStatus) {
            statusMatch = taskStatus === status;
          }

          if (taskStatus && sanitizedText) return textMatch && statusMatch;

          if (taskStatus) return statusMatch;

          if (sanitizedText) return textMatch;
        }),
      ]);
    },
    [tasksData]
  );

  console.log(tasksData, tasksView);

  const value = React.useMemo<TaskContextValues>(
    () => ({
      tasksData,
      tasksView,
      statuses: STATUSES,
      isLoading,

      addTask,
      updateTask,
      deleteTask,
      filterTasks,
    }),
    [
      tasksData,
      tasksView,
      addTask,
      updateTask,
      deleteTask,
      filterTasks,
      isLoading,
    ]
  );

  return (
    <TasksContext.Provider value={value}>{children}</TasksContext.Provider>
  );
};
