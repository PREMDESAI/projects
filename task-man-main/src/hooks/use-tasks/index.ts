export { TasksProvider } from './context';
export { useTasks } from './useTasks';
export * from './types';
