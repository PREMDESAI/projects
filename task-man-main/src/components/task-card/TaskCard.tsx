import { TrashIcon } from '@radix-ui/react-icons';
import { Box, Flex, IconButton, Text } from '@radix-ui/themes';
import React from 'react';
import { useNavigate } from 'react-router-dom';

import { Task, useTasks } from '@/hooks/use-tasks';

interface TaskCardProps {
  task: Task;
}

interface TaskCardContentProps extends TaskCardProps {}

const TaskCardContent: React.FC<TaskCardContentProps> = ({ task }) => (
  <Flex
    direction='column'
    gap='1'
    style={{
      background: '#fff',
      padding: '16px',
      borderRadius: '4px',
      boxShadow: '0 1px 2px #ccc',
      cursor: 'move',
    }}
  >
    <Text style={{ fontSize: '12px', color: '#3E63DD', fontWeight: 'bold' }}>
      {task.id}
    </Text>
    <Text style={{ fontSize: '14px', color: '#555' }}>{task.title}</Text>
  </Flex>
);

export const TaskCard: React.FC<TaskCardProps> = ({ task }) => {
  const navigate = useNavigate();
  const { deleteTask } = useTasks();
  const [isDragging, setIsDragging] = React.useState(false);

  const onDragStart = (
    event: React.DragEvent<HTMLDivElement>,
    taskId: Task['id']
  ) => {
    event.dataTransfer.setData('taskId', taskId);
    setIsDragging(true);
  };

  return (
    <Box
      style={{
        position: 'relative',
        opacity: isDragging ? 0.2 : 1,
      }}
      draggable
      onDragStart={(e) => onDragStart(e, task.id)}
      onDragEnd={() => setIsDragging(false)}
      onClick={() => navigate(`/tasks/${task.id}`)}
    >
      <TaskCardContent task={task} />
      <IconButton
        variant='ghost'
        style={{ position: 'absolute', top: 16, right: 16, cursor: 'pointer' }}
        onClick={(event) => {
          event.stopPropagation();
          deleteTask(task.id);
        }}
      >
        <TrashIcon width='18' height='18' />
      </IconButton>
    </Box>
  );
};
