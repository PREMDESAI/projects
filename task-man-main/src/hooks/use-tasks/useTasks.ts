import React from 'react';

import { TasksContext } from './context';
import { TaskStatus } from './types';
import { debounce } from 'throttle-debounce';

export const useTasks = (status?: TaskStatus) => {
  const context = React.useContext(TasksContext);
  if (!context) {
    throw new Error('useTasks should be wrapped inside TasksProvider');
  }

  const { tasksView, filterTasks } = context;

  return {
    ...context,
    tasksView: status
      ? tasksView.filter((task) => task.status === status)
      : tasksView,
    filterTasks: debounce(500, filterTasks),
  };
};
