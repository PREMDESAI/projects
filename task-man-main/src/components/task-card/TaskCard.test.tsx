import { render, screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { describe, it } from 'vitest';

import { Task } from '@/hooks/use-tasks';
import { TaskCard } from './TaskCard';

const task: Task = {
  id: 'TM-1',
  status: 'todo',
  title: 'Test Task',
  description: 'A fresh task to achieve today',
};

describe('TaskCard', () => {
  it('renders', async () => {
    const { container } = render(
      <MemoryRouter>
        <TaskCard task={task} />
      </MemoryRouter>
    );
    expect(screen.getByText('TM-1')).toBeVisible();
    expect(screen.getByText('Test Task')).toBeVisible();
    expect(container).toMatchSnapshot();
  });
});
