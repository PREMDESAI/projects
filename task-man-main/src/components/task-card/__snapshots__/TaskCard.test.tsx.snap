// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TaskCard > renders 1`] = `
<div>
  <div
    class="rt-Box"
    draggable="true"
    style="position: relative; opacity: 1;"
  >
    <div
      class="rt-Flex rt-r-fd-column rt-r-gap-1"
      style="background: rgb(255, 255, 255); padding: 16px; border-radius: 4px; box-shadow: 0 1px 2px #ccc; cursor: move;"
    >
      <span
        class="rt-Text"
        style="font-size: 12px; color: rgb(62, 99, 221); font-weight: bold;"
      >
        TM-1
      </span>
      <span
        class="rt-Text"
        style="font-size: 14px; color: rgb(85, 85, 85);"
      >
        Test Task
      </span>
    </div>
    <button
      class="rt-reset rt-BaseButton rt-r-size-2 rt-variant-ghost rt-IconButton"
      data-accent-color=""
      style="position: absolute; top: 16px; right: 16px; cursor: pointer;"
    >
      <svg
        fill="none"
        height="18"
        viewBox="0 0 15 15"
        width="18"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M5.5 1C5.22386 1 5 1.22386 5 1.5C5 1.77614 5.22386 2 5.5 2H9.5C9.77614 2 10 1.77614 10 1.5C10 1.22386 9.77614 1 9.5 1H5.5ZM3 3.5C3 3.22386 3.22386 3 3.5 3H5H10H11.5C11.7761 3 12 3.22386 12 3.5C12 3.77614 11.7761 4 11.5 4H11V12C11 12.5523 10.5523 13 10 13H5C4.44772 13 4 12.5523 4 12V4L3.5 4C3.22386 4 3 3.77614 3 3.5ZM5 4H10V12H5V4Z"
          fill="currentColor"
          fill-rule="evenodd"
        />
      </svg>
    </button>
  </div>
</div>
`;
